src/lib.rs

```diff
@@ -0,0 +1,408 @@
+pub mod sunset_db {
+    use std::collections::HashMap;
+    use std::error;
+    use std::ffi::OsStr;
+    use std::fmt;
+    use std::fs::{read_dir, File, OpenOptions};
+    use std::io::{self, Read, Seek, SeekFrom, Write};
+    use std::mem::size_of;
+    use std::os::unix::prelude::OsStrExt;
+    use std::path::{Path, PathBuf};
+
+    pub type Index = HashMap<String, u64>;
+
+    const SEGMENT_EXT: &str = "segment";
+
+    // TODO: Switch to using an empty byte string as the tombstone?
+    const TOMBSTONE: u64 = 1u64 << 63;
+    const ENCODED_TOMBSTONE: [u8; size_of::<u64>()] = (TOMBSTONE).to_be_bytes();
+
+    // NOTE: This will hold the file open as long as `Segment` is in memory.
+    struct Segment {
+        id: u64,
+        file: File,
+        index: Index,
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct KeyNotFoundError;
+    impl fmt::Display for KeyNotFoundError {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "key not found")
+        }
+    }
+    impl error::Error for KeyNotFoundError {}
+
+    #[derive(Debug, Clone)]
+    pub struct ExceedsMaxSizeError;
+    impl fmt::Display for ExceedsMaxSizeError {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "exceeds max size")
+        }
+    }
+    impl error::Error for ExceedsMaxSizeError {}
+
+    #[derive(Debug, Clone)]
+    pub struct InvalidChecksumError;
+    impl fmt::Display for InvalidChecksumError {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "invalid checksum")
+        }
+    }
+    impl error::Error for InvalidChecksumError {}
+
+    // TODO: Is there a better way to handle errors other than boxing them?
```

Comment:
Once you define your error type with `thiserror` you'll return your own error type and define the conversions within that type (impl `From` the specific error). Here you'll only need to use `?`.